File: frontend/src/api/ai.ts
import { DEFAULT_API_TIMEOUT_MS } from '../lib/apiConfig'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ProviderName = 'deepseek' | 'openai' | 'xai' | 'ollama' | 'google'
export type RoleName = 'screener' | 'tactical' | 'fundamental' | 'strategist'
export type SignalAction = 'BUY' | 'SELL' | 'NEUTRAL' | 'VETO'

export type ProviderStatus = {
  name: ProviderName
  healthy: boolean
  model: string
  lastChecked: string
}

export type RoleConfig = {
  name: RoleName
  provider: ProviderName
  model: string
  systemPromptId: string
  temperature: number
  maxTokens: number
  weight: number
  enabled: boolean
  fallbackProvider: ProviderName | null
  fallbackModel: string | null
}

export type SystemPrompt = {
  id: string
  role: RoleName
  version: number
  content: string
  description: string
  isActive: boolean
  createdAt: string
}

export type RoleVerdict = {
  role: RoleName
  action: SignalAction
  confidence: number
  reasoning: string
  metrics: Record<string, number>
}

export type ConsensusDecision = {
  symbol: string
  timeframe: string
  finalAction: SignalAction
  finalConfidence: number
  verdicts: RoleVerdict[]
  reasoning: string
  vetoedBy: RoleName | null
  totalCostUsd: number
  totalLatencyMs: number
  createdAt: string
}

export type EvaluationRequest = {
  symbol: string
  timeframe?: string
  candles?: Record<string, unknown>[]
  indicators?: Record<string, unknown>
  portfolio?: Record<string, unknown>
  riskLimits?: Record<string, unknown>
  roles?: RoleName[]
}

export type UsageRecord = {
  role: RoleName
  provider: ProviderName
  model: string
  tokensIn: number
  tokensOut: number
  costUsd: number
  latencyMs: number
  timestamp: string
  symbol: string
  success: boolean
}

export type UsageSummary = {
  totalCostUsd: number
  totalRequests: number
  byRole: Record<RoleName, { cost: number; requests: number; avgLatencyMs: number }>
  byProvider: Record<ProviderName, { cost: number; requests: number }>
}

export type DailyUsage = {
  date: string
  totalCostUsd: number
  totalRequests: number
  totalTokensIn: number
  totalTokensOut: number
}

export type BudgetStatus = {
  scope: string
  dailyLimit: number
  monthlyLimit: number
  dailySpent: number
  monthlySpent: number
  dailyExceeded: boolean
  monthlyExceeded: boolean
  exceeded: boolean
}

// ---------------------------------------------------------------------------
// API calls
// ---------------------------------------------------------------------------

/** Helper for fetch with abort timeout. */
async function fetchWithTimeout(
  url: string,
  init: RequestInit = {},
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<Response> {
  const controller = new AbortController()
  const timeoutId = setTimeout(() => controller.abort(), timeoutMs)
  try {
    const response = await fetch(url, { ...init, signal: controller.signal })
    return response
  } finally {
    clearTimeout(timeoutId)
  }
}

/**
 * Fetch all provider health statuses.
 */
export async function fetchProviderStatus(
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<ProviderStatus[]> {
  const response = await fetchWithTimeout('/api/ai/providers', {}, timeoutMs)
  if (!response.ok) throw new Error(`Failed to fetch providers: ${response.status}`)
  return await response.json()
}

/**
 * Fetch all role configurations.
 */
export async function fetchRoleConfigs(
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<RoleConfig[]> {
  const response = await fetchWithTimeout('/api/ai/roles', {}, timeoutMs)
  if (!response.ok) throw new Error(`Failed to fetch roles: ${response.status}`)
  return await response.json()
}

/**
 * Update a role configuration.
 */
export async function updateRoleConfig(
  role: RoleName,
  config: Partial<RoleConfig>,
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<RoleConfig> {
  const response = await fetchWithTimeout(`/api/ai/roles/${role}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(config),
  }, timeoutMs)
  if (!response.ok) throw new Error(`Failed to update role: ${response.status}`)
  return await response.json()
}

/**
 * Fetch all prompt versions for a role.
 */
export async function fetchPrompts(
  role: RoleName,
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<SystemPrompt[]> {
  const response = await fetchWithTimeout(`/api/ai/prompts/${role}`, {}, timeoutMs)
  if (!response.ok) throw new Error(`Failed to fetch prompts: ${response.status}`)
  return await response.json()
}

/**
 * Create a new prompt version.
 */
export async function createPrompt(
  prompt: Omit<SystemPrompt, 'createdAt'>,
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<SystemPrompt> {
  const response = await fetchWithTimeout('/api/ai/prompts', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(prompt),
  }, timeoutMs)
  if (!response.ok) throw new Error(`Failed to create prompt: ${response.status}`)
  return await response.json()
}

/**
 * Activate a specific prompt version.
 */
export async function activatePrompt(
  promptId: string,
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<{ status: string }> {
  const response = await fetchWithTimeout(`/api/ai/prompts/${promptId}/activate`, {
    method: 'PUT',
  }, timeoutMs)
  if (!response.ok) throw new Error(`Failed to activate prompt: ${response.status}`)
  return await response.json()
}

/**
 * Request a Multi-Brain evaluation for a symbol (POST — canonical endpoint).
 */
export async function evaluateSymbol(
  request: EvaluationRequest,
  timeoutMs: number = 120_000, // longer timeout for multi-agent
): Promise<ConsensusDecision> {
  const response = await fetchWithTimeout('/api/ai/evaluate', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request),
  }, timeoutMs)
  if (!response.ok) {
    const detail = await response.text().catch(() => '')
    throw new Error(`Evaluation failed (${response.status}): ${detail}`)
  }
  return await response.json()
}

/**
 * Fetch recent AI decisions.
 */
export async function fetchDecisions(
  params?: { symbol?: string; action?: string; limit?: number },
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<ConsensusDecision[]> {
  const searchParams = new URLSearchParams()
  if (params?.symbol) searchParams.set('symbol', params.symbol)
  if (params?.action) searchParams.set('action', params.action)
  if (params?.limit) searchParams.set('limit', String(params.limit))
  const qs = searchParams.toString()
  const url = `/api/ai/decisions${qs ? `?${qs}` : ''}`
  const response = await fetchWithTimeout(url, {}, timeoutMs)
  if (!response.ok) throw new Error(`Failed to fetch decisions: ${response.status}`)
  return await response.json()
}

/**
 * Fetch usage/cost summary.
 */
export async function fetchUsageSummary(
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<UsageSummary> {
  const response = await fetchWithTimeout('/api/ai/usage', {}, timeoutMs)
  if (!response.ok) throw new Error(`Failed to fetch usage: ${response.status}`)
  return await response.json()
}

/**
 * Fetch daily usage breakdown.
 */
export async function fetchDailyUsage(
  days: number = 30,
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<DailyUsage[]> {
  const response = await fetchWithTimeout(`/api/ai/usage/daily?days=${days}`, {}, timeoutMs)
  if (!response.ok) throw new Error(`Failed to fetch daily usage: ${response.status}`)
  return await response.json()
}

/**
 * Fetch budget status for a scope (global or role name).
 */
export async function fetchBudgetStatus(
  scope: string = 'global',
  timeoutMs: number = DEFAULT_API_TIMEOUT_MS,
): Promise<BudgetStatus> {
  const response = await fetchWithTimeout(`/api/ai/budget/status?scope=${scope}`, {}, timeoutMs)
  if (!response.ok) throw new Error(`Failed to fetch budget status: ${response.status}`)
  return await response.json()
}

File: db/models/ai.py
"""SQLAlchemy models for Multi-Brain AI tables.

These models map to the tables created by db/migrations/001_ai_tables.sql:
- system_prompts
- ai_role_configs
- ai_usage_log
- ai_decisions
"""

from __future__ import annotations


from sqlalchemy import BigInteger, Boolean, Column, Float, Integer, Text, ForeignKey, Index, DateTime, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import DeclarativeBase


class Base(DeclarativeBase):
    """Base class for all SQLAlchemy models."""

    pass


class SystemPrompt(Base):
    """Versioned system prompt storage per role.

    Table: system_prompts
    """

    __tablename__ = "system_prompts"

    id = Column(Text, primary_key=True)  # e.g. "tactical_v1"
    role = Column(Text, nullable=False)  # screener|tactical|fundamental|strategist
    version = Column(Integer, nullable=False, default=1)
    content = Column(Text, nullable=False)
    description = Column(Text, nullable=False, default="")
    is_active = Column(Boolean, nullable=False, default=True)
    created_at = Column(DateTime(timezone=True), nullable=False, server_default=func.now())

    __table_args__ = (Index("idx_system_prompts_role", "role", "is_active"),)

    def __repr__(self) -> str:
        return f"<SystemPrompt(id={self.id}, role={self.role}, v{self.version}, active={self.is_active})>"


class AIRoleConfig(Base):
    """Provider/model assignment per role.

    Table: ai_role_configs
    """

    __tablename__ = "ai_role_configs"

    name = Column(Text, primary_key=True)  # screener|tactical|fundamental|strategist
    provider = Column(Text, nullable=False)  # deepseek|openai|xai|ollama|google
    model = Column(Text, nullable=False)
    system_prompt_id = Column(Text, ForeignKey("system_prompts.id"), nullable=True)
    temperature = Column(Float, nullable=False, default=0.0)
    max_tokens = Column(Integer, nullable=False, default=4096)
    weight = Column(Float, nullable=False, default=1.0)
    enabled = Column(Boolean, nullable=False, default=True)
    fallback_provider = Column(Text, nullable=True)
    fallback_model = Column(Text, nullable=True)
    updated_at = Column(DateTime(timezone=True), nullable=False, server_default=func.now())

    def __repr__(self) -> str:
        return f"<AIRoleConfig(name={self.name}, provider={self.provider}, model={self.model})>"


class AIUsageLog(Base):
    """Cost/token tracking per AI request.

    Table: ai_usage_log
    """

    __tablename__ = "ai_usage_log"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    role = Column(Text, nullable=False)
    provider = Column(Text, nullable=False)
    model = Column(Text, nullable=False)
    tokens_in = Column(Integer, nullable=False, default=0)
    tokens_out = Column(Integer, nullable=False, default=0)
    cost_usd = Column(Float, nullable=False, default=0.0)
    latency_ms = Column(Float, nullable=False, default=0.0)
    symbol = Column(Text, nullable=False, default="")
    success = Column(Boolean, nullable=False, default=True)
    error = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), nullable=False, server_default=func.now())

    __table_args__ = (
        Index("idx_ai_usage_log_created", "created_at"),
        Index("idx_ai_usage_log_role", "role"),
        Index("idx_ai_usage_log_symbol", "symbol"),
    )

    def __repr__(self) -> str:
        return f"<AIUsageLog(id={self.id}, role={self.role}, cost=${self.cost_usd:.4f})>"


class AIDecision(Base):
    """Consensus decisions audit trail.

    Table: ai_decisions
    """

    __tablename__ = "ai_decisions"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    symbol = Column(Text, nullable=False)
    timeframe = Column(Text, nullable=False)
    final_action = Column(Text, nullable=False)  # BUY|SELL|NEUTRAL|VETO
    final_confidence = Column(Float, nullable=False, default=0.0)
    verdicts = Column(JSONB, nullable=False, default=lambda: [])  # list of RoleVerdict dicts
    reasoning = Column(Text, nullable=False, default="")
    vetoed_by = Column(Text, nullable=True)
    total_cost_usd = Column(Float, nullable=False, default=0.0)
    total_latency_ms = Column(Float, nullable=False, default=0.0)
    created_at = Column(DateTime(timezone=True), nullable=False, server_default=func.now())

    __table_args__ = (
        Index("idx_ai_decisions_symbol", "symbol", "created_at"),
        Index("idx_ai_decisions_action", "final_action"),
    )

    def __repr__(self) -> str:
        return f"<AIDecision(id={self.id}, symbol={self.symbol}, action={self.final_action})>"
